### PR TITLE
[SPARK-24889][Core] Update block info when unpersist rdds

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

We will update block info coming from executors, at the timing like caching a RDD. However, when removing RDDs with unpersisting, we don't ask to update block info. So the block info is not updated.

We can fix this with few options:

1. Ask to update block info when unpersisting

This is simplest but changes driver-executor communication a bit.

2. Update block info when processing the event of unpersisting RDD

We send a `SparkListenerUnpersistRDD` event when unpersisting RDD. When processing this event, we can update block info of the RDD. This only changes event processing code so the risk seems to be lower.

Currently this patch takes option 2 for lower risk. If we agree first option has no risk, we can change to it.

## How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/storage/RDDInfo.scala [hunk 1]
```diff
@@ -55,7 +55,7 @@ class RDDInfo(
 }
 
 private[spark] object RDDInfo {
-  private val callsiteForm = SparkEnv.get.conf.get(EVENT_LOG_CALLSITE_FORM)
+  private lazy val callsiteForm = SparkEnv.get.conf.get(EVENT_LOG_CALLSITE_FORM)
```
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/storage/RDDInfo.scala]: Is this related to the problem?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/storage/RDDInfo.scala]: I ran the test locally and this causes error when initializing `RDDInfo`. Actually I think this should be lazy because it is not always needed.

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/storage/RDDInfo.scala]: Thanks for explaining.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 2]
```diff
@@ -646,7 +647,47 @@ private[spark] class AppStatusListener(
...
+          exec.diskUsed = math.max(0, exec.diskUsed - rddDist.diskUsed)
+          executorsToUpdate += exec
+        }
+      }
+
+      // Use RDD partition info to update executor block info.
+      val partitions = liveRDD.getPartitions()
+
+      partitions.foreach { case (_, part) =>
+        val executors = part.executors
```
[the remark this answers is not in the document]
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: No, a partition can exist on more than one executor.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 3]
```diff
@@ -23,6 +23,7 @@ import java.util.function.Function
 
 import scala.collection.JavaConverters._
 import scala.collection.mutable.HashMap
+import scala.collection.mutable.HashSet
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: nit: merge with above import

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 4]
```diff
@@ -646,7 +647,47 @@ private[spark] class AppStatusListener(
   }
 
   override def onUnpersistRDD(event: SparkListenerUnpersistRDD): Unit = {
-    liveRDDs.remove(event.rddId)
+    liveRDDs.remove(event.rddId).foreach { liveRDD =>
+      val executorsToUpdate = new HashSet[LiveExecutor]()
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I'd just call `maybeUpdate` directly when updating the executor. It's cheaper than inserting into a set, since a duplicate call will just compare the timestamp of the last update and then do nothing.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 5]
```diff
@@ -646,7 +647,47 @@ private[spark] class AppStatusListener(
...
+      val distributions = liveRDD.getDistributions()
+
+      // Use RDD distribution to update executor memory and disk usage info.
+      distributions.foreach { case (executorId, rddDist) =>
+        val maybeExec = liveExecutors.get(executorId)
+
+        maybeExec.foreach { exec =>
+          if (exec.hasMemoryInfo) {
+            if (storageLevel.useOffHeap) {
+              exec.usedOffHeap = math.max(0, exec.usedOffHeap - rddDist.offHeapUsed)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I'd take the `newValue` method inside `updateRDDBlock` and make it a proper private method (with a better name), since then it becomes clearer why this logic is needed.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 6]
```diff
@@ -646,7 +647,47 @@ private[spark] class AppStatusListener(
...
   override def onUnpersistRDD(event: SparkListenerUnpersistRDD): Unit = {
-    liveRDDs.remove(event.rddId)
+    liveRDDs.remove(event.rddId).foreach { liveRDD =>
+      val executorsToUpdate = new HashSet[LiveExecutor]()
+      val storageLevel = liveRDD.info.storageLevel
+      val distributions = liveRDD.getDistributions()
+
+      // Use RDD distribution to update executor memory and disk usage info.
+      distributions.foreach { case (executorId, rddDist) =>
+        val maybeExec = liveExecutors.get(executorId)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: You don't need the variable, right? `liveExecutors.get(executorId).foreach` is enough?

(Same with `distributions` and `partitions` and `executors`.)

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 7]
```diff
@@ -538,6 +538,14 @@ private class LiveRDD(val info: RDDInfo) extends LiveEntity {
     distributions.get(exec.executorId)
   }
 
+  def getPartitions(): Map[String, LiveRDDPartition] = {
+    partitions.toMap
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: This makes a copy, no? Is there a need to make that copy?

**Review by vanzin — COMMENTED** [review]
Just some minor things.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 7 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: I think this just creates an immutable map and inserts all elements into it and so no copy of elements.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 6 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Yeah, right.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 7 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: But it makes a copy of the map, and that seems unnecessary. Of course it does not make a copy of the elements - it doesn't even know how to do that.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Do we prefer just to simply return the mutable HashMap?

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 4 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: An executor can be updated first when processing distributions and later when processing partitions. We want a duplicate call to do update in fact. If we just call `maybeUpdate` directly, the duplicate call will do nothing and miss updates from processing partitions.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala (hunk 7 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: Sure, it's an internal API. Listener code needs to avoid doing unnecessary things like copying stuff to avoid issues with dropping events.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 4 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Right. But it would be nice to avoid the hash set if possible. The less stuff listeners have to do, the better.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Ok. Good point for me. I updated the change to remove the hash set now.

**Comment by viirya** [comment]
@vanzin Thanks for the good review! I've updated this to address them all.

**Comment by viirya** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/status/LiveEntity.scala [hunk 8]
```diff
@@ -538,6 +538,10 @@ private class LiveRDD(val info: RDDInfo) extends LiveEntity {
     distributions.get(exec.executorId)
   }
 
+  def getPartitions(): scala.collection.Map[String, LiveRDDPartition] = partitions
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/status/LiveEntity.scala]: I'd have just exposed the fields but this is fine too.

**Comment by vanzin** [comment]
Merging to master, 2.4 and 2.3.

**Comment by vanzin** [comment]
There was a trivial conflict in 2.3, I fixed it manually.
